RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitProducer_gpu.cc

```diff
@@ -0,0 +1,136 @@
+      std::unique_ptr<EcalUncalibRecHitWorkerBaseClass> tmw(EcalUncalibRecHitFillDescriptionWorkerFactory::get()->create(itInfos->name_));
+      s = (itInfos->name_ >> edm::ParameterDescription<edm::ParameterSetDescription>("algoPSet", tmw->getAlgoDescription(), true));      
+    }
+    for (++itInfos; itInfos != infos.end(); ++itInfos) {
+      std::unique_ptr<EcalUncalibRecHitWorkerBaseClass> tmw(EcalUncalibRecHitFillDescriptionWorkerFactory::get()->create(itInfos->name_));
+      s = (std::move(s) or itInfos->name_ >> edm::ParameterDescription<edm::ParameterSetDescription>("algoPSet", tmw->getAlgoDescription(), true));
+    }
+    desc.ifValue(edm::ParameterDescription<std::string>("algo", "EcalUncalibRecHitWorkerMultiFitGPU", true), std::move(s));
+    
+    descriptions.addDefault(desc);
+  }
+
+  for (std::vector<edmplugin::PluginInfo>::const_iterator itInfos = infos.begin(); itInfos != infos.end(); itInfos++) {
+    std::unique_ptr<EcalUncalibRecHitWorkerBaseClass> fdWorker(EcalUncalibRecHitFillDescriptionWorkerFactory::get()->create(itInfos->name_)); 
+
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("EBdigiCollection", edm::InputTag("ecalDigis","ebDigis"));
+    desc.add<std::string>("EEhitCollection", "EcalUncalibRecHitsEE");
+    desc.add<edm::InputTag>("EEdigiCollection", edm::InputTag("ecalDigis","eeDigis"));
+    desc.add<std::string>("EBhitCollection", "EcalUncalibRecHitsEB");
+    desc.add<std::string>("algo", itInfos->name_);
+    desc.add<edm::ParameterSetDescription>("algoPSet", fdWorker->getAlgoDescription()); 
+    
+    std::string algoName = itInfos->name_.substr(itInfos->name_.find("Worker")+6, itInfos->name_.length());
+    descriptions.add("ecal"+algoName+"UncalibRecHitGPU", desc);
+//    descriptions.addDefault(desc);
+  }
+
+}
+
+void
+EcalUncalibRecHitProducerGPU::produce(edm::Event& evt, const edm::EventSetup& es) {
+
+        using namespace edm;
+
+        Handle< EBDigiCollection > pEBDigis;
+        Handle< EEDigiCollection > pEEDigis;
+
+        const EBDigiCollection* ebDigis =nullptr;
+        const EEDigiCollection* eeDigis =nullptr;
+
+
+	evt.getByToken( ebDigiCollectionToken_, pEBDigis);		
+	ebDigis = pEBDigis.product(); // get a ptr to the produc
+	edm::LogInfo("EcalUncalibRecHitInfo") << "total # ebDigis: " << ebDigis->size() ;
+                    
+	evt.getByToken( eeDigiCollectionToken_, pEEDigis);            
+	eeDigis = pEEDigis.product(); // get a ptr to the product
+	edm::LogInfo("EcalUncalibRecHitInfo") << "total # eeDigis: " << eeDigis->size() ;
+        
+
+        // tranparently get things from event setup
+        worker_->set(es);
+        worker_->set(evt);
+
+        // prepare output
+        auto ebUncalibRechits = std::make_unique<EBUncalibratedRecHitCollection>();
+        auto eeUncalibRechits = std::make_unique<EEUncalibratedRecHitCollection>();
+
+        // loop over EB digis
+        if (ebDigis)
+            worker_->run(evt, *ebDigis, *ebUncalibRechits);
```

Comment:
The original CPU code (where this class seems to be copy-pasted) should be rewritten... This `if` is useless as `ebDigis` is always non-`nullptr` (in absence of product framework will throw on line 106). 

In addition the complex helper plugin mechanism in `fillDescriptions()` could likely be replaced with https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He